docs/proposals/20210222-azure-machinepool-machine.md

```diff
@@ -0,0 +1,332 @@
+    // can be added as events to the MachinePool object and/or logged in the
+    // controller's output.
+    // +optional
+    FailureMessage *string `json:"failureMessage,omitempty"`
+    
+    // Conditions defines current service state of the AzureMachinePool.
+    // +optional
+    Conditions clusterv1.Conditions `json:"conditions,omitempty"`
+    
+    // LongRunningOperationState saves the state for an Azure long running operations so it can be continued on the
+    // next reconciliation loop.
+    // +optional
+    LongRunningOperationState *infrav1.Future `json:"longRunningOperationState,omitempty"`
+    
+    // LatestModelApplied indicates the instance is running the most up-to-date VMSS model. A VMSS model describes
+    // the image version the VM is running. If the instance is not running the latest model, it means the instance
+    // may not be running the version of Kubernetes the Machine Pool has specified and needs to be updated.
+    LatestModelApplied bool `json:"latestModelApplied"`
+    
+    // Ready is true when the provider resource is ready.
+    // +optional
+    Ready bool `json:"ready"`
+}
+```
+
+#### Proposed Controller Changes
+
+* Create a new AzureMachinePoolMachine controller.
+* Remove VMSS instance status tracking logic from AzureMachinePool controller and moving it to AzureMachinePoolMachine
+  controller.
+* (optional) Introduce rate limiting behavior to AzureMachinePool* controllers to ensure Azure API limits are not 
+  exceeded.  
+
+#### Proposed Changes of Responsibility
+Currently in CAPZ, the AzureMachinePool controller is responsible for both the Virtual Machine Scale Set (VMSS) and the
+instances created by the VMSS. The proposed change would separate the responsibility of managing the state of the VMSS
+and the instances created by the VMSS. This would introduce a new AzureMachinePoolMachine controller and a new
+MachinePoolMachineScope. The responsibilities would be as follows.
+
+**AzureMachinePool Responsibilities:**
+- Create AzureMachinePoolMachine instances when a new VMSS instance is observed. The AzureMachinePoolMachine spec should
+  have the `ProviderID` field set with the observed resource ID. The AzureMachinePool should also be added to the
+  AzureMachinePoolMachine's OwnerReferences. 
+- Upgrade / replace selection of AzureMachinePoolMachine instances. When a change to the AzureMachinePool model occurs,
+  the MachinePoolScope will be responsible for coordinating the rollout of the updated model by selecting
+  AzureMachinePoolMachines to delete / replace with respect to MaxUnavailable.
+- Scale up. AzureMachinePool should increase the number of VMSS replicas if the replica count increases on MachinePool
+- Scale down. AzureMachinePool should select and delete AzureMachinePoolMachines that are overprovisioned with respect
+  to MaxUnavailable.
+- Clean up. When a AzureMachinePoolMachine is no longer in the list of instances in Azure, but a matching
```

Comment:
This implies Azure is the source of truth and the goal state. But at the same time, we're saying you can trigger a VMSS VM delete by deleting an AzureMachinePoolMachine which kind of goes against that assumption. IMO the CRs should be the source of truth / goal state, and the controller's job should be to reconcile against that goal state and try to get the real world (Azure resources) to match. Now if something happens out of band to delete a VMSS instance, the controller should see that and then our options are either 1) recreate it (that's what we do for all resources excepts VMs right now) or, and this is probably the option we need to pick if we're going to follow the immutable machines principle, 2) mark the AzureMachinePoolMachine as "Failed" and have some external remediation mechanism (equivalent to a MachineHealthCheck) who deletes failed machines - which should cause a new instance to be created to replace it since the desired replica count will become > than the actual number of replicas.

Reply:
Imagine when we over-provision VMs. Azure will spin up multiple, wait for the first to successfully deploy, then delete the others that didn't win the provisioning race. In this case, the loser VMs will be deleted by Azure. We must clean up the AMPMs which may have been created while the provisioning race was running.

Reply:
That makes sense